### PR TITLE
Support set/get xattr

### PR DESCRIPTION
**Pull request by Jackson-Wang-7**
### What changes are proposed in this pull request?

Supports setXattr API of Alluxio FileSystem, it will set the corresponding attributes to UFS. If UFS is a filesystem, it's set through the setXattr interface of UFS. if UFS is object storage, it's set through setTagging API.

### Why are the changes needed?
Based on the 3.x architecture, recover support for the setXattr interface. There are still many places where the interface is used, and it should not be supported at present.

### CODE REVIEW

File: dora/core/common/src/main/java/alluxio/util/io/FileUtils.java [hunk 1]
```diff
@@ -98,6 +102,67 @@ public static void changeLocalFilePermission(String filePath, String perms) {
...
+   * Changes local file's user defined file attribute.
+   * @param filePath the file to change attribute
+   * @param name the name of attribute
+   * @param value the value of attribute
+   */
+  public static void changeLocalFileUserDefinedAttribute(String filePath, String name,
+                                                         byte[] value) {
+    try {
+      UserDefinedFileAttributeView attributeView =
+          Files.getFileAttributeView(Paths.get(filePath), UserDefinedFileAttributeView.class);
```
Comment (elega) [inline dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:115]: can you add a comment to explain what is an attribute view what is this for?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -439,9 +440,12 @@ protected int getCachedPercentage(alluxio.grpc.FileInfo fi, String ufsFullPath)
    *
    * @param status
    * @param ufsFullPath
+   * @param xattrMap
    * @return a FileInfo
    */
-  public alluxio.grpc.FileInfo buildFileInfoFromUfsStatus(UfsStatus status, String ufsFullPath) {
+  public alluxio.grpc.FileInfo buildFileInfoFromUfsStatus(UfsStatus status, String ufsFullPath,
```
Comment (elega) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:446]: nit: add a nullable annotation here

File: dora/core/common/src/main/java/alluxio/util/io/FileUtils.java (hunk 1 above)
Comment (elega) [reply dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:115]: is the attr only set locally and will lose after a worker restarts?

Comment (Jackson-Wang-7) [reply dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:115]: > is the attr only set locally and will lose after a worker restarts?
No, it sets the attribute to the local file system, so it will not change after a worker restarts. (works in CentOS, not work in MacOS)

File: dora/underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 3]
```diff
@@ -130,6 +132,16 @@ public String getUnderFSType() {
   @Override
   public void setOwner(String path, String user, String group) {}
 
+  @Override
+  public void setObjectTagging(String path, String name, String value) throws IOException {
+    throw new UnimplementedRuntimeException("set attribute is not implemented");
```
Comment (jiacheliu3) [inline dora/underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java:137]:
```suggestion
    throw new UnimplementedRuntimeException("setAttribute is not implemented in GCS V1, use V2 instead");
```

File: dora/underfs/gcs/src/main/java/alluxio/underfs/gcs/v2/GCSV2UnderFileSystem.java [hunk 4]
```diff
@@ -137,6 +139,32 @@ public PositionReader openPositionRead(String path, long fileLength) {
   @Override
   public void setOwner(String path, String user, String group) {}
 
+  @Override
+  public void setObjectTagging(String path, String name, String value) throws IOException {
+    Map<String, String> newMetadata = new HashMap<>();
+    newMetadata.put(name, value);
```
Comment (jiacheliu3) [inline dora/underfs/gcs/src/main/java/alluxio/underfs/gcs/v2/GCSV2UnderFileSystem.java:145]: this actually creates a Map of 16 instances. What about just `ImmutableMap.of(name, value)`?

File: dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 5]
```diff
@@ -743,6 +746,39 @@ public void setOwner(String path, String user, String group) throws IOException
     }
   }
 
+  @Override
+  public void setAttribute(String path, String name, byte[] value) throws IOException {
+    if (StringUtils.isEmpty(name)) {
+      return;
```
Comment (jiacheliu3) [inline dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java:752]: I guess `name` is the inode name like `file.txt`? So when someone trys to set with name "" we can print a warning or something?

File: dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java [hunk 6]
```diff
@@ -30,4 +34,20 @@ public void testWriteMultiBlockFileTest() throws Exception {
...
+    OutputStream os = mUfs.create(testFilePath, getCreateOption());
+    os.close();
+    assertEquals(0, mUfs.getStatus(testFilePath).asUfsFileStatus().getContentLength());
+
+    // setXAttr
+    String attrKey = "key1";
+    String attrValue = "value1";
+    mUfs.setAttribute(testFilePath, attrKey, attrValue.getBytes());
+    Map<String, String> attrMap =  mUfs.getAttribute(testFilePath);
+    assertEquals(attrMap.get(attrKey), attrValue);
```
Comment (jiacheliu3) [inline dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java:51]: i figure some other UT cases can be added on corner cases at least, like unsetting an attribute, setting attr twice and observe a successful overwrite etc.

File: dora/underfs/obs/src/main/java/alluxio/underfs/obs/OBSUnderFileSystem.java [hunk 7]
```diff
@@ -188,6 +191,29 @@ public PositionReader openPositionRead(String path, long fileLength) {
   public void setOwner(String path, String user, String group) {
   }
 
+  @Override
+  public void setObjectTagging(String path, String name, String value) throws IOException {
+    ObjectMetadata metadata = mClient.getObjectMetadata(mBucketName, path);
+    SetObjectMetadataRequest request = new SetObjectMetadataRequest(mBucketName, path);
```
Comment (jiacheliu3) [inline dora/underfs/obs/src/main/java/alluxio/underfs/obs/OBSUnderFileSystem.java:197]: mention the same race condition in code comments

File: dora/underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 8]
```diff
@@ -191,6 +194,21 @@ public PositionReader openPositionRead(String path, long fileLength) {
   @Override
   public void setOwner(String path, String user, String group) {}
 
+  @Override
+  public void setObjectTagging(String path, String name, String value) throws IOException {
+    TagSet taggingResult = mClient.getObjectTagging(mBucketName, path);
```
Comment (jiacheliu3) [inline dora/underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java:199]: similarly on the race condition

File: dora/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 9]
```diff
@@ -485,6 +492,40 @@ public String getUnderFSType() {
   @Override
   public void setOwner(String path, String user, String group) {}
 
+  @Override
+  public void setObjectTagging(String path, String name, String value) throws IOException {
+    GetObjectTaggingRequest getTaggingReq = new GetObjectTaggingRequest(mBucketName, path);
```
Comment (jiacheliu3) [inline dora/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:497]: similarly on the race condition

File: dora/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 10]
```diff
@@ -485,6 +492,40 @@ public String getUnderFSType() {
...
+    GetObjectTaggingResult taggingResult = mClient.getObjectTagging(getTaggingReq);
+    List<Tag> tagList = taggingResult.getTagSet();
+    boolean existed = false;
```
Comment (jiacheliu3) [inline dora/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:500]:
```suggestion
    boolean matchFound = false;
```

File: dora/core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 11]
```diff
@@ -862,6 +862,25 @@ default Optional<UfsStatus[]> listStatuses(String path, ListOptions options) thr
...
+  void setAttribute(String path, String name, byte[] value) throws IOException;
+
+  /**
+   * Gets the attribute of the given path. An empty implementation should be provided if
+   * unsupported.
+   *
+   * @param path the path of the file
+   * @return the map of all attributes
+   */
+  Map<String, String> getAttribute(String path) throws IOException;
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:882]: i feel we can call  this `getAttributes()`

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks a lot for the work! I left some comments PTAL

File: dora/core/server/proxy/src/main/java/alluxio/proxy/s3/S3BucketTask.java [hunk 12]
```diff
@@ -214,7 +214,13 @@ public Response continueTask() {
           try {
             List<URIStatus> children = mHandler.getMetaFS().listStatus(new AlluxioURI(
                     S3RestUtils.MULTIPART_UPLOADS_METADATA_DIR));
-            return ListMultipartUploadsResult.buildFromStatuses(bucket, children);
+            // GetListing doesn't contain the xattr info, so get status for MPU files.
+            List<URIStatus> mpuList = new ArrayList<>();
+            for (URIStatus status : children) {
+              URIStatus filemeta = mHandler.getMetaFS().getStatus(new AlluxioURI(status.getPath()));
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply dora/core/server/proxy/src/main/java/alluxio/proxy/s3/S3BucketTask.java:220]: as long as other calls don't pay this price, i'm good with this part

File: dora/core/common/src/main/java/alluxio/underfs/UnderFileSystemWithLogging.java [hunk 13]
```diff
@@ -1172,17 +1173,17 @@ public String methodName() {
 
       @Override
       public String toString() {
-        return String.format("path=%s, name=%s, value=%s", path, name, value);
+        return String.format("path=%s, name=%s, value=%s", path, name, Arrays.toString(value));
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/underfs/UnderFileSystemWithLogging.java:1176]: nit: we have helpers you can use, it's fine as-is here but you have more options next time
```
return MoreObjects.toStringHelper(this).add(xxx,xxx).toString();
```

File: dora/core/common/src/main/java/alluxio/util/io/FileUtils.java [hunk 14]
```diff
@@ -141,7 +143,7 @@ public static Map<String, String> getLocalFileUserDefinedAttribute(String filePa
       if (attributeView == null) {
         throw new UnimplementedRuntimeException("set attribute is not implemented");
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:144]:
```
        throw new UnimplementedRuntimeException(
            "UFS does not support getting FileAttributeView with Java.");
```

File: dora/core/common/src/main/java/alluxio/util/io/FileUtils.java [hunk 15]
```diff
@@ -98,6 +102,67 @@ public static void changeLocalFilePermission(String filePath, String perms) {
...
+    try {
+      UserDefinedFileAttributeView attributeView =
+          Files.getFileAttributeView(Paths.get(filePath), UserDefinedFileAttributeView.class);
+      if (attributeView == null) {
+        throw new UnimplementedRuntimeException("set attribute is not implemented");
+      }
+      Map<String, String> attrMap = new HashMap<>();
+      for (String attributeName : attributeView.list()) {
+        int attributeSize = attributeView.size(attributeName);
+        ByteBuffer attributeBuffer = ByteBuffer.allocate(attributeSize);
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:149]: sanity check ^

File: dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 16]
```diff
@@ -749,6 +749,7 @@ public void setOwner(String path, String user, String group) throws IOException
   @Override
   public void setAttribute(String path, String name, byte[] value) throws IOException {
     if (StringUtils.isEmpty(name)) {
+      LOG.warn("Try to set Xattr to an empty file name.");
```
Comment (jiacheliu3) [inline dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java:752]: actually is the `name` here file name or user name?
```suggestion
      LOG.warn("Try to set Xattr to an empty file name for path {}", path);
```

File: dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java [hunk 17]
```diff
@@ -47,7 +47,35 @@ public void testSetAndGetXAttribute() throws Exception {
...
+    // setXAttr
+    String attrKey = "key1";
+    String attrValue1 = "value1";
+    String attrValue2 = "value2";
+    mUfs.setAttribute(testFilePath, attrKey, attrValue1.getBytes());
+    mUfs.setAttribute(testFilePath, attrKey, attrValue2.getBytes());
+    Map<String, String> attrMap =  mUfs.getAttributes(testFilePath);
+    assertEquals(attrMap.size(), 1);
+    assertEquals(attrMap.get(attrKey), attrValue2);
+    mUfs.deleteFile(testFilePath);
```
Comment (jiacheliu3) [inline dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java:79]: delete at the end of the test is not guaranted to happen. Could you move that to `@After`?

**Review by jiacheliu3 — APPROVED** [review]
Mostly LGTM with some relatively minor comments. Good to go when you see fit, and I'm stamping ahead of time. Thanks a lot for the work!

We might be able to add more UT cases, but you can do that in follow up PRs (and not necessarily but you alone)

File: dora/core/common/src/main/java/alluxio/util/io/FileUtils.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Jackson-Wang-7) [reply dora/core/common/src/main/java/alluxio/util/io/FileUtils.java:149]: I used `ByteBuffer.allocate` here instead of `ByteBuffer.allocateDirect`, so the buffer is managed by JVM. it will be automatically destroyed and recycled after GC.

File: dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java (hunk 17 above)
Comment (Jackson-Wang-7) [reply dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/hdfs3/HdfsUnderFileSystemIntegrationTest.java:79]: I will move it to the `finally` part. I'm not sure if deletion is necessary. I haven't seen it in other test cases, I just follow my personal habits.

**Comment by yyongycy** [comment]
What is the max size of tagging?

**Comment by Jackson-Wang-7** [comment]
> What is the max size of tagging?

we have the configuration property to limit the tag number we can set, the max size is 10.

**Comment by Jackson-Wang-7** [comment]
alluxio-bot, merge this please

**Comment by Jackson-Wang-7** [comment]
alluxio-bot, merge this please
